sx126x/sx126x.go

```diff
@@ -0,0 +1,690 @@
+	p[2] = uint8((timeoutRtcStep >> 0) & 0xFF)
+	d.ExecSetCommand(SX126X_CMD_SET_TX, p[:])
+}
+
+// SetRx() sets the device in RX mode
+// timeout is expressed in RTC Step unit (15uS)
+// Value of 0x000000 => No timeout. Rx Single mode.
+// Value of 0xffffff => Rx Continuous mode
+// Other values => Timeout active. The device remains in RX until countdown or packet received
+func (d *Device) SetRx(timeoutRtcStep uint32) {
+	var p [3]uint8
+	p[0] = uint8(((timeoutRtcStep >> 16) & 0xFF))
+	p[1] = uint8(((timeoutRtcStep >> 8) & 0xFF))
+	p[2] = uint8(((timeoutRtcStep >> 0) & 0xFF))
+	d.ExecSetCommand(SX126X_CMD_SET_RX, p[:])
+}
+
+// StopTimerOnPreamble allows the user to select if the timer is stopped upon preamble detection of SyncWord / header detection.
+func (d *Device) StopTimerOnPreamble(enable bool) {
+	var p [1]uint8
+	if enable {
+		p[0] = 1
+	} else {
+		p[0] = 0
+	}
+	d.ExecSetCommand(SX126X_CMD_STOP_TIMER_ON_PREAMBLE, p[:])
+}
+
+// SetRegulatorMode sets the regulator more (depends on hardware implementation)
+func (d *Device) SetRegulatorMode(mode uint8) {
+	p := []uint8{mode}
+	d.ExecSetCommand(SX126X_CMD_SET_REGULATOR_MODE, p[:])
+}
+
+// Calibrate starts the calibration of a block defined by calibParam
+func (d *Device) Calibrate(calibParam uint8) {
+	p := []uint8{calibParam}
+	d.ExecSetCommand(SX126X_CMD_CALIBRATE, p[:])
+}
+
+// CalibrateImage calibrates the image rejection of the device for the device operating
+func (d *Device) CalibrateImage(freq uint32) {
+	var calFreq [2]uint8
+	if freq > 900000000 {
+		calFreq[0] = 0xE1
+		calFreq[1] = 0xE9
+	} else if freq > 850000000 {
+		calFreq[0] = 0xD7
+		calFreq[1] = 0xD8
+	} else if freq > 770000000 {
+		calFreq[0] = 0xC1
+		calFreq[1] = 0xC5
+	} else if freq > 460000000 {
+		calFreq[0] = 0x75
+		calFreq[1] = 0x81
+	} else if freq > 425000000 {
+		calFreq[0] = 0x6B
+		calFreq[1] = 0x6F
+	}
+	d.ExecSetCommand(SX126X_CMD_CALIBRATE_IMAGE, calFreq[:])
+}
+
+// SetPaConfig sets the Power Amplifier configuration (R)
+func (d *Device) SetPaConfig(paDutyCycle, hpMax, deviceSel, paLut uint8) {
+	var p [4]uint8
+	p[0] = paDutyCycle
+	p[1] = hpMax
+	p[2] = deviceSel
+	p[3] = paLut
+	d.ExecSetCommand(SX126X_CMD_SET_PA_CONFIG, p[:])
+}
+
+// SetRxTxFallbackMode defines into which mode the chip goes after a successful transmission or after a packet reception.
+func (d *Device) SetRxTxFallbackMode(fallbackMode uint8) {
+	d.ExecSetCommand(SX126X_CMD_SET_RX_TX_FALLBACK_MODE, []uint8{fallbackMode})
+}
+
+// --------------------------------------------------
+// Registers and Buffers
+// --------------------------------------------------
+
+// ReadRegister reads register value
+func (d *Device) ReadRegister(addr, size uint16) ([]uint8, error) {
+	d.CheckDeviceReady()
+	d.SpiSetNss(false)
+	// Send command
+	d.SpiTx(SX126X_CMD_READ_REGISTER)
+	d.SpiTx(uint8((addr & 0xFF00) >> 8))
+	d.SpiTx(uint8(addr & 0x00FF))
+	d.SpiTx(uint8(0x00)) // This byte is for status (unused yet)
+
+	var ret []uint8
```

Comment:
You can define this as `var ret [size]unit8` to avoid heap allocations, and then instead of the loop use something like this:

```go
d.spi.Tx(nil, ret[:])
```

Reply:
```
var ret [size]uint8 fails with 
```

... fails with : 
```
array length size (variable of type uint16) must be constant
```
I guess we should use slice instead. 
I could also declare it as ret[MAX_READ_SIZE]uint8, or something like that. 

But I'm not sure that still makes sense.

Reply:
> I could also declare it as a static array with some max size (eg: ret[MAX_READ_SIZE]uint8), but would that make sense ?

Yes that would be better. The idea is to avoid all of the heap allocations by reusing an array as a slice hence avoiding use of `make` when not really needed.

Reply:
Thanks @deadprogram 

Do you think I could use such commands:  
```
tinygo build -print-allocs=. -target gnse -o test.hex  2>&1 | grep heap  | grep sx126x
/home/olivier/go/src/tinygo.org/x/drivers/sx126x/sx126x.go:293:14: object allocated on the heap: escapes at line 294
/home/olivier/go/src/tinygo.org/x/drivers/sx126x/spi_stm32wl.go:20:22: object allocated on the heap: escapes at line 20
/home/olivier/go/src/tinygo.org/x/drivers/sx126x/spi_stm32wl.go:16:2: object allocated on the heap: escapes at line 26
```

... to improve my code ? 
Because it doesn't report heap allocation on the line you spotted earlier. 
```
253	var ret []uint8
```

Can this be an potential issue at runtime ?

Thanks.

Reply:
That is a question best answered by @aykevl

Reply:
At the end, I defined a spi_buffer[256]  (in the device struct) . 
256 is the max allowed size for reading datas from the radio.

Please tell me if it's good or not  (My mind is not yet very clear on stack/heap usage) 

Thx